pkg/sources/filesystem/filesystem.go

```diff
@@ -229,3 +230,43 @@ func (s *Source) Enumerate(ctx context.Context, units chan<- sources.Enumeration
 	}
 	return nil
 }
+
+// ChunkUnit implements SourceUnitChunker interface.
+func (s *Source) ChunkUnit(ctx context.Context, unit sources.SourceUnit, reporter sources.ChunkReporter) error {
+	path := unit.SourceUnitID()
+	logger := ctx.Logger().WithValues("path", path)
+
+	cleanPath := filepath.Clean(path)
+	fileInfo, err := os.Stat(cleanPath)
+	if err != nil {
+		return reporter.ChunkErr(ctx, fmt.Errorf("unable to get file info: %w", err))
+	}
+
+	ch := make(chan *sources.Chunk)
+	go func() {
+		defer close(ch)
+		if fileInfo.IsDir() {
+			// TODO: Finer grain error tracking of individual chunks.
+			err = s.scanDir(ctx, cleanPath, ch)
+		} else {
+			// TODO: Finer grain error tracking of individual
+			// chunks (in the case of archives).
+			err = s.scanFile(ctx, cleanPath, ch)
+		}
+	}()
+
+	for chunk := range ch {
+		if chunk == nil {
+			continue
+		}
+		if err := reporter.ChunkOk(ctx, *chunk); err != nil {
```

Comment:
What does it mean if there's an error here? Is terminating the chunking process the correct thing to do if it happens?

Reply:
Any time the callbacks return an error it should be propagated up. Usually this will be a context cancellation.

Reply:
Yes! That was the intention with the design. Any error returned by those callbacks should terminate chunking.